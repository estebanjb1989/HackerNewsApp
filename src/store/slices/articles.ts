import { createSlice, createAsyncThunk, current } from "@reduxjs/toolkit";
import notifee from '@notifee/react-native';
import { IArticle, IState } from "@interfaces";

interface IArticleState {
  articles: IArticle[];
  deletedArticles: string[];
  favoriteArticles: string[];
  loading: boolean;
  query: string;
}

const initialState = {
  articles: [],
  deletedArticles: [],
  favoriteArticles: [],
  loading: false,
  query: "tech"
} as IArticleState;

const articlesSlice = createSlice({
  name: "articlesSlice",
  initialState,
  reducers: {
    setLoading(state, action) {
      state.loading = action.payload
    },
    setArticles(state, action) {
      state.articles = action.payload;
    },
    deleteArticle(state, action) {
      state.deletedArticles = [...(state.deletedArticles || []), action.payload]
    },
    addToFavorites(state, action) {
      state.favoriteArticles = [...(state.favoriteArticles || []), action.payload]
    },
    removeFromFavorites(state, action) {
      state.favoriteArticles = state.favoriteArticles?.filter(fa => fa !== action.payload)
    },
    setQuery(state, action) {
      state.query = action.payload
    }
  },
});

// Define an async action creator
export const fetchArticles = createAsyncThunk(
  'articles/fetch',
  async (args, { dispatch, getState }) => {
    try {
      dispatch(setLoading(true))
      const query = (getState() as IState).articles?.query

      const response = await fetch("https://hn.algolia.com/api/v1/search_by_date?query=" + query)
      const data = await response.json()

      const articles = data.hits
        .filter((hit: any) => hit.story_title && hit.story_url)
        .map((hit: any) => ({
          id: hit.created_at_i,
          title: hit.story_title,
          author: hit.author,
          createdAt: hit.created_at,
          url: hit.story_url
        } as IArticle))
        .sort((a: any, b: any) => {
          if (a.created_at_i > b.created_at_i) return -1;
          if (a.created_at_i < b.created_at_i) return 1;
          return 0;
        })

      const notificationsEnabled = (getState() as IState)?.settings?.notificationsEnabled
      const currentArticles = (getState() as IState)?.articles?.articles

      if (notificationsEnabled && currentArticles?.length > 0) {
        const newArticles = articles.filter((art: IArticle) =>
          !currentArticles?.some((curr: IArticle) => curr.id === art.id)
        )

        const channelId = await notifee.createChannel({
          id: 'default',
          name: 'Default Channel',
        });

        for (const newArticle of newArticles) {
          await notifee.displayNotification({
            title: "New article by " + newArticle.author,
            body: newArticle.title,
            data: {
              url: newArticle.url || ""
            },
            android: {
              channelId,
              smallIcon: 'ic_launcher', // optional, defaults to 'ic_launcher'.
              // pressAction is needed if you want the notification to open the app when pressed
              pressAction: {
                id: 'default',
              },
            },
          });
        }
      }

      return articles
    } catch (err) {
      console.error(err)
      throw err
    } finally {
      dispatch(setLoading(false))
    }
  }
);

export const {
  setLoading,
  setArticles,
  deleteArticle,
  addToFavorites,
  removeFromFavorites,
  setQuery
} = articlesSlice.actions;

export default articlesSlice.reducer;
