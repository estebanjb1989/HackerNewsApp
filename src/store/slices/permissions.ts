import { createSlice } from "@reduxjs/toolkit";

interface IPermissionsState {
  notificationsPermissionStatus: number | null;
}

const initialState = {
  notificationsPermissionStatus: null,
} as IPermissionsState;

const permissionsSlice = createSlice({
  name: "permissionsSlice",
  initialState,
  reducers: {
    setNotificationsPermissionStatus(state, action) {
      state.notificationsPermissionStatus = action.payload
    },    
  },
});

export const {
  setNotificationsPermissionStatus
} = permissionsSlice.actions;

export default permissionsSlice.reducer;
