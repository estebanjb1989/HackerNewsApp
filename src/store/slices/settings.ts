import { createSlice } from "@reduxjs/toolkit";

interface ISettingsState {
  notificationsEnabled: boolean;
}

const initialState = {
  notificationsEnabled: false,
} as ISettingsState;

const settingsSlice = createSlice({
  name: "settingsSlice",
  initialState,
  reducers: {
    setNotificationsEnabled(state, action) {
      state.notificationsEnabled = action.payload
    },    
  },
});

export const {
  setNotificationsEnabled
} = settingsSlice.actions;

export default settingsSlice.reducer;
