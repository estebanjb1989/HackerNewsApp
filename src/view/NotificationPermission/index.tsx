import React from 'react';
import {
    StyleSheet,
    Text,
    TouchableOpacity,
    View,    
    Alert
} from 'react-native';
import notifee, { AuthorizationStatus }  from '@notifee/react-native';
import { Title } from "@component/Title"
import { DrawerNavigationProp } from "@react-navigation/drawer";
import { ParamListBase, useNavigation } from '@react-navigation/native';
import { useDispatch } from 'react-redux';
import { setNotificationsEnabled } from '@store/slices/settings';
import { setNotificationsPermissionStatus } from '@store/slices/permissions';

function NotificationPermission() {
    const navigation = useNavigation<DrawerNavigationProp<ParamListBase>>()
    const dispatch = useDispatch()

    const handleContinue = async () => {        
        try {            
            const settings = await notifee.requestPermission();
            const status = settings?.authorizationStatus                   
            dispatch(setNotificationsPermissionStatus(status))            
            if (status === AuthorizationStatus.AUTHORIZED || status === AuthorizationStatus.PROVISIONAL) {
                dispatch(setNotificationsEnabled(true))
                navigation.navigate("App")
            } else if (status === AuthorizationStatus.DENIED) {
                Alert.alert("Alert", "Permissions have been denied, please try again or enable them in the OS settings")                
                dispatch(setNotificationsEnabled(false))
                console.log('Notification permission is denied');
            }            
        } catch (err) {
            dispatch(setNotificationsEnabled(false))
            console.error(err)
        }
    }

    const handleDeny = () => {
        dispatch(setNotificationsEnabled(false))
        navigation.navigate("App")
    }

    return (
        <View style={styles.outerContainer}>
            <View style={styles.container}>
                <View />
                <View>
                    <Title text="Hacker News App" centered />
                </View>
                <View style={styles.middleContainer}>
                    <Text style={styles.text}>We request your permission to receive notifications in order to let you know when articles of your interest are created</Text>
                    <TouchableOpacity testID="notification-permission_continue" onPress={handleContinue}>
                        <View style={styles.button}>
                            <Text style={styles.buttonText}>Continue</Text>
                        </View>
                    </TouchableOpacity>
                </View>
                <View>
                    <TouchableOpacity testID="notification-permission_deny"  onPress={handleDeny}>
                        <View style={styles.button}>
                            <Text style={styles.buttonText}>Don't send me notifications please</Text>
                        </View>
                    </TouchableOpacity>
                </View>
                <View />
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    outerContainer: {
        backgroundColor: "lightgrey",
        flex: 1
    },
    container: {
        backgroundColor: "white",
        justifyContent: "space-between",
        alignItems: "center",
        flex: 1,
        margin: 12,
        borderWidth: 1,
        borderColor: "lightgrey",
        borderRadius: 6
    },
    middleContainer: {
        alignItems: "center"
    },
    text: {
        textAlign: "center",
        width: 240
    },
    buttonText: {
        textAlign: "center",
        color: "white",
        minWidth: 120
    },
    button: {

        backgroundColor: "blue",
        borderRadius: 6,
        marginTop: 12,
        padding: 12,
        alignItems: "center",
        justifyContent: "center"
    }
});

export default NotificationPermission;
