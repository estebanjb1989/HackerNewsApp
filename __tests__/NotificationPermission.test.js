import React from 'react';
import { render, fireEvent } from '@testing-library/react-native'
import { useSelector } from "react-redux"
import NotificationPermission from "@view/NotificationPermission"

import * as permissionsSlice from "@store/slices/permissions"
import * as settingsSlice from "@store/slices/settings"

const mockState = {
  permissions: {
    notificationsPermissionStatus: null
  }
}

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

const spySetNotificationsPermissionStatus = jest.spyOn(permissionsSlice,
  "setNotificationsPermissionStatus");

const spySetNotificationsEnabled = jest.spyOn(settingsSlice,
  "setNotificationsEnabled"
)

jest.mock('@notifee/react-native', () => ({
  requestPermission: jest.fn(),
  AuthorizationStatus: jest.fn()
}))

jest.mock("@store/slices/permissions", () => ({
  setNotificationsPermissionStatus: jest.fn(),
}))

jest.mock("@store/slices/settings", () => ({
  setNotificationsEnabled: jest.fn(),
}))

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn()
}))

describe("Notification Permission view", () => {
  beforeEach(() => {
    useSelector.mockImplementation((cb) => cb(mockState))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const tree = render(<NotificationPermission />);
    expect(tree).toMatchSnapshot();
  });

  it('should call set notifications enabled true and navigate to app if user decides to continue', async () => {
    const tree = render(<NotificationPermission />);
    const continueButton = await tree.findByTestId("notification-permission_continue")
    fireEvent.press(continueButton)
    expect(spySetNotificationsPermissionStatus).toHaveBeenCalled()
  });

  fit('should call set notifications enabled false and navigate to app if user decides to deny', async () => {
    const tree = render(<NotificationPermission />);
    const denyButton = await tree.findByTestId("notification-permission_deny")
    fireEvent.press(denyButton)    
    expect(spySetNotificationsEnabled).toHaveBeenCalled()
  });
})

