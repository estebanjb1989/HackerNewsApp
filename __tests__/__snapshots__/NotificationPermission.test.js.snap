// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Permission view renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "lightgrey",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "lightgrey",
        "borderRadius": 6,
        "borderWidth": 1,
        "flex": 1,
        "justifyContent": "space-between",
        "margin": 12,
      }
    }
  >
    <View />
    <View>
      <Text
        style={
          [
            {
              "fontSize": 20,
              "marginHorizontal": 8,
              "marginVertical": 8,
              "minWidth": 240,
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Hacker News App
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          {
            "textAlign": "center",
            "width": 240,
          }
        }
      >
        We request your permission to receive notifications in order to let you know when articles of your interest are created
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="notification-permission_continue"
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "blue",
              "borderRadius": 6,
              "justifyContent": "center",
              "marginTop": 12,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "minWidth": 120,
                "textAlign": "center",
              }
            }
          >
            Continue
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="notification-permission_deny"
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "blue",
              "borderRadius": 6,
              "justifyContent": "center",
              "marginTop": 12,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "minWidth": 120,
                "textAlign": "center",
              }
            }
          >
            Don't send me notifications please
          </Text>
        </View>
      </View>
    </View>
    <View />
  </View>
</View>
`;

exports[`Notification Permission view should call set notifications enabled true and navigate to app if user decides to continue 1`] = `
<View
  style={
    {
      "backgroundColor": "lightgrey",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "lightgrey",
        "borderRadius": 6,
        "borderWidth": 1,
        "flex": 1,
        "justifyContent": "space-between",
        "margin": 12,
      }
    }
  >
    <View />
    <View>
      <Text
        style={
          [
            {
              "fontSize": 20,
              "marginHorizontal": 8,
              "marginVertical": 8,
              "minWidth": 240,
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Hacker News App
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          {
            "textAlign": "center",
            "width": 240,
          }
        }
      >
        We request your permission to receive notifications in order to let you know when articles of your interest are created
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "blue",
              "borderRadius": 6,
              "justifyContent": "center",
              "marginTop": 12,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "minWidth": 120,
                "textAlign": "center",
              }
            }
          >
            Continue
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "blue",
              "borderRadius": 6,
              "justifyContent": "center",
              "marginTop": 12,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "minWidth": 120,
                "textAlign": "center",
              }
            }
          >
            Don't send me notifications please
          </Text>
        </View>
      </View>
    </View>
    <View />
  </View>
</View>
`;
